widgetry/src/mapspace/world.rs

```diff
@@ -0,0 +1,485 @@
+    }
+
+    /// Draw a tooltip while hovering over this object.
+    pub fn tooltip(mut self, txt: Text) -> Self {
+        assert!(self.tooltip.is_none(), "already specified tooltip");
+        // TODO Or should this implicitly mark the object as hoverable? Is it weird to base this
+        // off drawing?
+        assert!(
+            self.draw_hover.is_some(),
+            "first specify how to draw hovered"
+        );
+        self.tooltip = Some(txt);
+        self
+    }
+
+    /// Mark the object as clickable. `WorldOutcome::ClickedObject` will be fired.
+    pub fn clickable(mut self) -> Self {
+        assert!(!self.clickable, "called clickable twice");
+        self.clickable = true;
+        self
+    }
+
+    /// Mark the object as draggable. The user can hover on this object, then click and drag it.
+    /// `WorldOutcome::Dragging` events will be fired.
+    ///
+    /// Note that dragging an object doesn't transform it at all (for example, by translating its
+    /// hitbox). The caller is responsible for doing that.
+    pub fn draggable(mut self) -> Self {
+        assert!(!self.draggable, "called draggable twice");
+        self.draggable = true;
+        self
+    }
+
+    /// While the user hovers over this object, they can press a key to perform the specified
+    /// action. `WorldOutcome::Keypress` will be fired.
+    pub fn hotkey<I: Into<MultiKey>>(mut self, key: I, action: &'static str) -> Self {
+        // TODO Check for duplicate keybindings
+        self.keybindings.push((key.into(), action));
+        self
+    }
+
+    /// Finalize the object, adding it to the `World`.
+    pub fn build(mut self, ctx: &mut EventCtx) {
+        let hitbox = self.hitbox.take().expect("didn't specify hitbox");
+        let bounds = hitbox.get_bounds();
+        let quadtree_id = self
+            .world
+            .quadtree
+            .insert_with_box(self.id, bounds.as_bbox());
+
+        self.world.objects.insert(
+            self.id,
+            Object {
+                _id: self.id,
+                _quadtree_id: quadtree_id,
+                hitbox,
+                zorder: self.zorder,
+                draw_normal: self
+                    .draw_normal
+                    .expect("didn't specify how to draw normally")
+                    .build(ctx),
+                draw_hover: self.draw_hover.take().map(|draw| draw.build(ctx)),
+                tooltip: self.tooltip,
+                clickable: self.clickable,
+                draggable: self.draggable,
+                keybindings: self.keybindings,
+            },
+        );
+    }
+}
+
+struct Object<ID: ObjectID> {
+    _id: ID,
+    _quadtree_id: ItemId,
+    hitbox: Polygon,
+    zorder: usize,
+    draw_normal: ToggleZoomed,
+    draw_hover: Option<ToggleZoomed>,
+    tooltip: Option<Text>,
+    clickable: bool,
+    draggable: bool,
+    // TODO How should we communicate these keypresses are possible? Something standard, like
```

Comment:
For buttons, we have a standard style for showing the hotkey. Maybe we could do something similar here, once we decide what that UI pattern should be. Many places in the traffic sim use that "on-screen display" bottom bar thing, but that's something I came up with way way back, and has never been Yuwen-approved. ;)